x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/IndexServiceAccountsTokenStore.java

```diff
@@ -0,0 +1,196 @@
+    }
+
+    @Override
+    void doAuthenticate(ServiceAccountToken token, ActionListener<Boolean> listener) {
+        final GetRequest getRequest = client
+            .prepareGet(SECURITY_MAIN_ALIAS, docIdForToken(token))
+            .setFetchSource(true)
+            .request();
+        securityIndex.checkIndexVersionThenExecute(listener::onFailure, () ->
+            executeAsyncWithOrigin(client, SECURITY_ORIGIN, GetAction.INSTANCE, getRequest, ActionListener.<GetResponse>wrap(response -> {
+                if (response.isExists()) {
+                    final String tokenHash = (String) response.getSource().get("password");
+                    assert tokenHash != null : "service account token hash cannot be null";
+                    listener.onResponse(Hasher.verifyHash(token.getSecret(), tokenHash.toCharArray()));
+                } else {
+                    logger.trace("service account token [{}] not found in index", token.getQualifiedName());
+                    listener.onResponse(false);
+                }}, listener::onFailure)));
+    }
+
+    public void createToken(Authentication authentication, CreateServiceAccountTokenRequest request,
+                            ActionListener<CreateServiceAccountTokenResponse> listener) {
+        final ServiceAccountId accountId = new ServiceAccountId(request.getNamespace(), request.getServiceName());
+        if (false == ServiceAccountService.isServiceAccountPrincipal(accountId.asPrincipal())) {
+            listener.onFailure(new IllegalArgumentException("service account [" + accountId + "] does not exist"));
+            return;
+        }
+        final ServiceAccountToken token = ServiceAccountToken.newToken(accountId, request.getTokenName());
+        try (XContentBuilder builder = newDocument(authentication, token)) {
+            final IndexRequest indexRequest =
+                client.prepareIndex(SECURITY_MAIN_ALIAS)
+                    .setId(docIdForToken(token))
+                    .setSource(builder)
+                    .setOpType(OpType.CREATE)
+                    .setRefreshPolicy(request.getRefreshPolicy())
+                    .request();
+            final BulkRequest bulkRequest = toSingleItemBulkRequest(indexRequest);
+
+            securityIndex.prepareIndexIfNeededThenExecute(listener::onFailure, () -> {
+                executeAsyncWithOrigin(client, SECURITY_ORIGIN, BulkAction.INSTANCE, bulkRequest,
+                    TransportSingleItemBulkWriteAction.<IndexResponse>wrapBulkResponse(ActionListener.wrap(response -> {
+                        if (DocWriteResponse.Result.CREATED == response.getResult()) {
+                            listener.onResponse(CreateServiceAccountTokenResponse.created(
+                                token.getTokenName(), token.asBearerString()));
+                        } else {
+                            listener.onResponse(CreateServiceAccountTokenResponse.notCreated());
```

Comment:
This seems weird. What does it mean? How do we have an `OpType.CREATE` but then get  a result that is not `CREATED` ?

I suspect we don't need `notCreated` to even exists. The `created: true` result was proposed when I thought we might support overwriting credentials, but since we're not current planning that, I don't think we ever have a `created:false` version.

Reply:
Ah good catch. These are left over code before I add `OpType.CREATE`. Will adjust. Thanks